fdbserver/storageserver.actor.cpp

```diff
@@ -6594,27 +6588,48 @@ ACTOR Future<Standalone<VectorRef<BlobGranuleChunkRef>>> tryReadBlobGranules(Tra
 	}
 }
 
-// Read keys from blob storage if they exist. Fail back to tryGetRange, which reads keys
-// from storage servers with locally attached disks
+// Read blob granules metadata. The key range can cross tenant bundary.
+ACTOR Future<Standalone<VectorRef<BlobGranuleChunkRef>>> readBlobGranuleChunks(Transaction* tr,
+                                                                               Database cx,
+                                                                               KeyRangeRef keys,
+                                                                               Version fetchVersion) {
+	loop {
+		state Standalone<VectorRef<BlobGranuleChunkRef>> results;
```

Comment:
I was thinking about weird potential races and i think there is one here.
If for some strange reason the fetch version was ahead of the read version we read the blobbified ranges at, you could potentially miss a newly blobbified range and thus its data?..
This is tricky as listBlobbifiedRanges doesn't provide the read version it got the ranges at (or it can read them from multiple actually).
I think as long as you could guarantee that the version that the ranges were read at was >= fetch version you wouldn't miss anything, but then you'd also have to handle the case where the range didn't exist yet at the fetch version and would get blob_granule_transaction_too_old.
I'm not sure we have a great solution here though, I'll need to think about it more

Reply:
Actually I think we need to handler retrying the fetch at a later version anyway if you get any sort of "too old", like storage servers already do, so this is probably totally fine

Reply:
> you could guarantee that the version that the ranges were read at was >= fetch version you wouldn't miss anything
For full cluster restore, it should be guaranteed. We load blobified ranges from manifest and advance version to max of blob granule version, then in a few mins we start fetch keys. When it comes to here and we call listBlobifiedRanges, the read version should be able to make sure we can read all blobified ranges. 

It's even safer that we retry on blob_granule_transaction_too_old errors. So I think it should be all right. I didn't see related errors from 100K correctness test. I can run probably 1m correctness test later to ensure we are good with this kind of edge cases